test/sync_tests/conflicting.test.js

```diff
@@ -0,0 +1,763 @@
+    let note = await Factory.createMappedNote(this.application);
+    await this.application.itemManager.setItemDirty(note.uuid);
+    await this.application.syncService.sync(syncOptions);
+    this.expectedItemCount++;
+
+    const newTitle = `${Math.random()}`;
+    /** First modify the item without saving so that
+     * our local contents digress from the server's */
+    await this.application.changeItem(note.uuid, (mutator) => {
+      mutator.title = `${Math.random()}`;
+    });
+    await this.application.itemManager.changeItem(note.uuid, (mutator) => {
+      // modify this item to have stale values
+      mutator.title = newTitle;
+      mutator.updated_at = Factory.yesterday();
+    });
+
+    // We expect this item to be duplicated
+    this.expectedItemCount++;
+
+    await this.application.syncService.clearSyncPositionTokens();
+    await this.application.syncService.sync(syncOptions);
+
+    note = this.application.findItem(note.uuid);
+    // We expect the item title to be the new title, and not rolled back to original value
+    expect(note.content.title).to.equal(newTitle);
+
+    const allItems = this.application.itemManager.items;
+    expect(allItems.length).to.equal(this.expectedItemCount);
+  });
+
+  it('should handle sync conflicts by not duplicating same data', async function () {
+    const note = await Factory.createMappedNote(this.application);
+    this.expectedItemCount++;
+    await this.application.itemManager.setItemDirty(note.uuid);
+    await this.application.syncService.sync(syncOptions);
+
+    // keep item as is and set dirty
+    await this.application.itemManager.setItemDirty(note.uuid);
+
+    // clear sync token so that all items are retrieved on next sync
+    this.application.syncService.clearSyncPositionTokens();
+
+    await this.application.syncService.sync(syncOptions);
+    expect(this.application.itemManager.items.length).to.equal(
+      this.expectedItemCount
+    );
+  });
+
+  it('clearing conflict_of on two clients simultaneously should keep us in sync', async function () {
+    const note = await Factory.createMappedNote(this.application);
+    await this.application.itemManager.setItemDirty(note.uuid);
+    this.expectedItemCount++;
+
+    await this.application.changeAndSaveItem(
+      note.uuid,
+      (mutator) => {
+        // client A
+        mutator.content.conflict_of = 'foo';
+      },
+      undefined,
+      undefined,
+      syncOptions
+    );
+
+    // client B
+    await this.application.syncService.clearSyncPositionTokens();
+    await this.application.itemManager.changeItem(
+      note.uuid,
+      (mutator) => {
+        mutator.content.conflict_of = 'bar';
+      },
+      undefined,
+      undefined,
+      syncOptions
+    );
+
+    // conflict_of is a key to ignore when comparing content, so item should
+    // not be duplicated.
+    await this.application.syncService.sync(syncOptions);
+  });
```

Comment:
Are you confident in using `afterEach` hooks for assertions that are core to an individual test, instead of just cleanup? It seems dangerous to me, but I've not explored this thought deeply and I trust your judgement. Just want to make sure this is your intent.

Reply:
Good point